[WebGPU] Instance being retained longer than its corresponding RemoteGPU

mwyrzykowski opened the pull request:
#### 1b1f0299a6918f7ab386fee8d5f442913385cc85
<pre>
[WebGPU] Instance being retained longer than its corresponding RemoteGPU
<a href="https://bugs.webkit.org/show_bug.cgi?id=279138">https://bugs.webkit.org/show_bug.cgi?id=279138</a>
<a href="https://rdar.apple.com/135292412">rdar://135292412</a>

Reviewed by Dan Glastonbury.

Chance Instance from Ref&lt;&gt; to a ThreadSafeWeakPtr&lt;&gt; to break a retain
cycle causing Instance to outlive it&apos;s corresponding RemoteGPU instance.

* Source/WebGPU/WebGPU/Adapter.h:
(WebGPU::Adapter::instance const):
* Source/WebGPU/WebGPU/Adapter.mm:
(WebGPU::Adapter::Adapter):
* Source/WebGPU/WebGPU/BindGroup.mm:
(WebGPU::Device::createExternalTextureFromPixelBuffer const):
* Source/WebGPU/WebGPU/Device.h:
(WebGPU::Device::instance const):
* Source/WebGPU/WebGPU/Device.mm:
(WebGPU::Device::Device):
(WebGPU::Device::setOwnerWithIdentity const):
(WebGPU::Device::popErrorScope):
(WebGPU::Device::webProcessID const):
* Source/WebGPU/WebGPU/ExternalTexture.mm:
(WebGPU::ExternalTexture::update):
* Source/WebGPU/WebGPU/Instance.h:
* Source/WebGPU/WebGPU/Queue.mm:
(WebGPU::Queue::scheduleWork):
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::Device::createRenderPipelineAsync):

Canonical link: <a href="https://commits.webkit.org/283190@main">https://commits.webkit.org/283190@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/92f08d5c51e88bd7146b69fbda7cb06fff4872f8

| Misc | iOS, visionOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/65456 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/44827 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/55/builds/18074 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/69482 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/59/builds/16065 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/52628 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/61/builds/16345 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/52565 "Passed tests") | [✅ 🧪 wincairo-tests](https://ews-build.webkit.org/#/builders/60/builds/11140 "Passed tests") 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/68523 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/41427 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/56656 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/33189 "Passed tests") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/42/builds/38100 "Passed tests") | | [✅ 🛠 wpe-cairo](https://ews-build.webkit.org/#/builders/65/builds/14941 "Built successfully") | 
| | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/59937 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/63/builds/14372 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/71187 "Built successfully") | 
| | [✅ 🛠 vision](https://ews-build.webkit.org/#/builders/87/builds/9410 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/62/builds/13833 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/59885 "Passed tests") | 
| | [✅ 🛠 vision-sim](https://ews-build.webkit.org/#/builders/86/builds/9442 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/56717 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/60158 "Passed tests") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/14439 "Built successfully and passed tests") | [✅ 🧪 vision-wk2](https://ews-build.webkit.org/#/builders/88/builds/7781 "Passed tests") | [✅ 🧪 mac-intel-wk2](https://ews-build.webkit.org/#/builders/119/builds/1429 "Passed tests") | | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/40637 "Built successfully") | | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/41713 "Built successfully") | | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/42896 "Built successfully") | | | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/41457 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

djg commented on Source/WebGPU/WebGPU/Device.mm:
```diff
@@ -163,12 +163,13 @@ static void captureFrame(id<MTLDevice> captureObject)
-                return;
-            protectedThis->loseTheDevice(WGPUDeviceLostReason_Undefined);
-        });
+        if (auto instance = protectedThis->instance(); instance.get()) {
```
In the other examples you appear to set the precedent of `if (auto inst = ...; inst.get())`

mwyrzykowski replied:
oh yes, I didn't like the abbreviation but it was to avoid the compiler error:
```
Variable 'instance' declared with deduced type 'auto' cannot appear in its own initializer
```
and I couldn't think of a better name. We avoid the abbreviation here because `this` is not in scope, but rather `protectedThis`

djg reviewed (approved):
:shipit:

mwyrzykowski committed:
[WebGPU] Instance being retained longer than its corresponding RemoteGPU
https://bugs.webkit.org/show_bug.cgi?id=279138
rdar://135292412

Reviewed by Dan Glastonbury.

Chance Instance from Ref<> to a ThreadSafeWeakPtr<> to break a retain
cycle causing Instance to outlive it's corresponding RemoteGPU instance.

* Source/WebGPU/WebGPU/Adapter.h:
(WebGPU::Adapter::instance const):
* Source/WebGPU/WebGPU/Adapter.mm:
(WebGPU::Adapter::Adapter):
* Source/WebGPU/WebGPU/BindGroup.mm:
(WebGPU::Device::createExternalTextureFromPixelBuffer const):
* Source/WebGPU/WebGPU/Device.h:
(WebGPU::Device::instance const):
* Source/WebGPU/WebGPU/Device.mm:
(WebGPU::Device::Device):
(WebGPU::Device::setOwnerWithIdentity const):
(WebGPU::Device::popErrorScope):
(WebGPU::Device::webProcessID const):
* Source/WebGPU/WebGPU/ExternalTexture.mm:
(WebGPU::ExternalTexture::update):
* Source/WebGPU/WebGPU/Instance.h:
* Source/WebGPU/WebGPU/Queue.mm:
(WebGPU::Queue::scheduleWork):
* Source/WebGPU/WebGPU/RenderPipeline.mm:
(WebGPU::Device::createRenderPipelineAsync):

Canonical link: https://commits.webkit.org/283190@main
